web/lua/opmserver/paginator.lua

```diff
@@ -0,0 +1,177 @@
+
+local _M = {}
+local mt = { __index = _M }
+
+local str_find = string.find
+local str_gsub = string.gsub
+local tab_concat = table.concat
+local abs = math.abs
+local modf = math.modf
+local ceil = math.ceil
+
+local function tab_append(t, v, idx)
+    idx = idx or #t
+    idx = idx + 1
+    t[idx] = v
+
+    return idx
+end
+
+function _M:new(url, total, page_size, page)
+
+    total = total or 1
+    page_size = page_size or 20
+    page = page or 1
+
+    local this = {
+        total = total,
+        page_size = page_size,
+        page = page,
+        page_count = ceil(total / page_size),
+        url = url
+    }
+    
+    setmetatable(this, mt)
+
+    return this
+end
+
+function _M:set_url(page)
+
+    local url = self.url
+
+    if str_find(url, '[&%?]page=') then
+        url = str_gsub(url, '([&%?])page=%d+', '%1page=' .. page)
+
+    else
+        if not str_find(url, '%?') then
+            url = url .. '?'
+        end
+
+        if str_find(url, '%w=') then
+            url = url .. '&'
+        end
+
+        url = url .. 'page=' .. page
+    end
+
+    return url
+end
+
+function _M:prev()
+
+    local page
+
+    if self.page == 1 then
+        page = 1
+    
+    else 
+        page = self.page - 1
+    end
```

Comment:
this logic (line 65 to 70) is confused.
we even don't need the `page` value when `self.page == 1`, right?